clang/test/SemaSYCL/free_function_kernel_params.cpp

```diff
@@ -0,0 +1,65 @@
+// RUN: %clang_cc1 -internal-isystem %S/Inputs -fsycl-is-device -ast-dump \
+// RUN: %s -o - | FileCheck %s
+// UNSUPPORTED: system-windows
+// This test checks parameter rewriting for free functions with parameters
+// of type scalar, pointer, simple struct and struct with pointers.
+// Windows support will be added later.
+
+#include "sycl.hpp"
+
+struct Simple {
+  int x;
+  char c[100];
+  float f;
+};
+
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void ff_2(int *ptr, int start, int end, struct Simple S) {
+  for (int i = start; i <= end; i++)
+    ptr[i] = start + S.x + S.f + S.c[2];
+}
+// CHECK: FunctionDecl {{.*}} __free_function_ff_2 'void (__global int *, int, int, struct Simple)'
+// CHECK-NEXT: ParmVarDecl {{.*}} _arg_ptr '__global int *'
+// CHECK-NEXT: ParmVarDecl {{.*}} _arg_start 'int'
+// CHECK-NEXT: ParmVarDecl {{.*}} _arg_end 'int'
+// CHECK-NEXT: ParmVarDecl {{.*}} _arg_S 'struct Simple':'Simple'
+// CHECK-NEXT: CompoundStmt
+// CHECK-NEXT: CallExpr {{.*}} 'void'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'void (*)(int *, int, int, struct Simple)' <FunctionToPointerDecay>
+// CHECK-NEXT: DeclRefExpr {{.*}} 'void (int *, int, int, struct Simple)' Function {{.*}} 'ff_2' 'void (int *, int, int, struct Simple)'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'int *' <AddressSpaceConversion>
```

Comment:
The AST checked in the lit test is generated by our compiler.

Reply:
My point is that what we generate with our code should match what Clang generates in a comparable call, and it currently does not. So modifications might be required in our BodyCreator code for free functions.

Reply:
Which clang are you referring to? Our clang generates what's in the test. For example:

call.cpp
```
__attribute__((sycl_device))
void ff_callee(int *ptr, int start, int end) {
  for (int i = start; i <= end; i++)
    ptr[i] = start + end;
}


__attribute__((sycl_device))
[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
  void ff_caller(int* ptr, int start, int end) {
  ff_callee(ptr, start, end);
}
``` 
when compiled as:
clang -cc1 -internal-isystem /iusers/rdeodhar/cmplr/freefunc/llvm/clang/test/SemaSYCL/Inputs -triple spir64-unknown-unknown -fsycl-is-device -ast-dump call.cpp -o -

generates:
```
 |-CompoundStmt 0x55ba006afe30 <<invalid sloc>>
  | `-CallExpr 0x55ba006afdf8 <col:8> 'void'
  |   |-ImplicitCastExpr 0x55ba006afde0 <col:8> 'void (*)(int *, int, int)' <FunctionToPointerDecay>
  |   | `-DeclRefExpr 0x55ba006afdc0 <col:8> 'void (int *, int, int)' Function 0x55ba006af5a8 'ff_caller' 'void (int *, int, int)'
  |   |-ImplicitCastExpr 0x55ba006afc68 <col:8> 'int *' <AddressSpaceConversion>
  |   | `-ImplicitCastExpr 0x55ba006afc50 <col:8> '__global int *' <LValueToRValue>
  |   |   `-DeclRefExpr 0x55ba006afc30 <col:8> '__global int *' lvalue ParmVar 0x55ba006afbc8 '_arg_ptr' '__global int *'
  |   |-DeclRefExpr 0x55ba006afd00 <col:8> 'int' ParmVar 0x55ba006afc98 '_arg_start' 'int'
  |   `-DeclRefExpr 0x55ba006afda0 <col:8> 'int' ParmVar 0x55ba006afd38 '_arg_end' 'int'
```

Reply:
I am talking about the AST generated for any function call using any clang. 

The way we usually go about figuring out what expressions we need to generate for kernel body/initializations/calls/etc is usually by looking at what clang already generates in a comparable call, and we try to match that. 

Here https://godbolt.org/z/de3d5qTso you will notice that Clang generates implicit lvalue to rvalue casts when handling a function call. These are missing in your lit test because your code does not generate them. My point is your code might need to generate these casts.

Reply:
I am not sure if those casts are necessary though which is why I've tagged @tahonermann  and @AaronBallman to confirm

Reply:
I would expect those casts to exist -- whenever you access an object (something with a name and an address, essentially), you need to obtain the value from it. That's an [lvalue to rvalue conversion](https://eel.is/c++draft/conv.lval). Parameters are lvalues, you can see that in one of the CHECK strings:
```
// CHECK-NEXT: DeclRefExpr {{.*}} '__global int *' lvalue ParmVar {{.*}} '_arg_ptr' '__global int *'
```
See `Sema::DefaultLvalueConversion()`.

(I'm surprised this test passes because the other `_arg_start` and `_arg_end` should be `'int' lvalue ParmVar`, it's worth figuring out why that's not the case if the test is passing).

Another thing I note is that the generated names are not quite right. We should not be generating a name like `_arg_ptr` because that's not a reserved identifier and users are free to have external identifiers or macros with that name, etc. It should be `__arg_ptr` so that it's using a reserved name.

Reply:
OK, I've made it look like you expect. DeclRefExpr using lvalue and then an LValuetoRValue cast.
I've also changed the names to use __arg_ a as prefix.
(Note: there is another instance of use of _arg_ as prefix, for accessors. However, since that is not part of my change I am leaving it alone. It does have a TODO comment on it though.)

Reply:
>> Another thing I note is that the generated names are not quite right. We should not be generating a name like _arg_ptr because that's not a reserved identifier and users are free to have external identifiers or macros with that name, etc. It should be __arg_ptr so that it's using a reserved name.

@AaronBallman this is an existing problem. Current kernel generation uses a single underscore. Should we deviate from it for free functions or keep it consistent and do a change for all of it later? I do not know how disruptive it would be for anyone depending of names we generate (though I would be surprised if this is the case)

Reply:
I don't insist on using double underscores, but my personal feeling is: why make the problem worse given that this is new code? Might as well do it right from the start so there's less chance of users being impacted (especially because it's unclear how long it will be before someone fixes the other issues).

Reply:
For the new code I am using an __arg prefix as that is the preferred naming.